utility/src/main/java/org/oppia/android/util/caching/testing/FakeAssetRepository.kt

```diff
@@ -66,17 +67,42 @@ class FakeAssetRepository @Inject constructor(
   }
 
   private fun loadTextFile(assetName: String): String {
-    return trackedAssets.computeIfAbsent(assetName) {
-      prodImpl.loadTextFileFromLocalAssets(assetName)
-    } as? String ?: error("Asset doesn't exist: $assetName")
+    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+      trackedAssets.computeIfAbsent(assetName) {
```

Comment:
Here & elsewhere: could we use Kotlin's https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/get-or-put.html instead? That would avoid the SDK-specific API that we're using here (and avoid needing to have two different implementations).

Reply:
Thanks for the suggestion , 
I think its enough to use getOrPut combined with synchronized function of kotlin to avoid unnecessary computation , as the main use of computeIfAbsent seemingly is to avoid unnecessary computation.

I think we can replace if condition by using code in the below format : 

```
return memoizedLocales[localeContext] ?: synchronized(memoizedLocales) {
      memoizedLocales.getOrPut(localeContext) {
        val chooser = profileChooserSelector.findBestChooser(localeContext)
        val primaryLocaleSource = LocaleSource.createFromPrimary(localeContext)
        val fallbackLocaleSource = LocaleSource.createFromFallback(localeContext)
        val proposal = chooser.findBestProposal(primaryLocaleSource, fallbackLocaleSource)
        proposal.computedLocale
      }
    }
```

instead of writing like : 

```
return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      memoizedLocales.computeIfAbsent(localeContext) {
        val chooser = profileChooserSelector.findBestChooser(localeContext)
        val primaryLocaleSource = LocaleSource.createFromPrimary(localeContext)
        val fallbackLocaleSource = LocaleSource.createFromFallback(localeContext)
        val proposal = chooser.findBestProposal(primaryLocaleSource, fallbackLocaleSource)
        return@computeIfAbsent proposal.computedLocale
      }
    } else {
      // Note : Using get/PutIfAbsent For API Level below 24 as computeIfAbsent is introduced in API Level 24
      val locale = memoizedLocales[localeContext] ?: synchronized(memoizedLocales) {
        val chooser = profileChooserSelector.findBestChooser(localeContext)
        val primaryLocaleSource = LocaleSource.createFromPrimary(localeContext)
        val fallbackLocaleSource = LocaleSource.createFromFallback(localeContext)
        val proposal = chooser.findBestProposal(primaryLocaleSource, fallbackLocaleSource)
        memoizedLocales.putIfAbsent(localeContext, proposal.computedLocale)
        proposal.computedLocale
      }
      return locale
    }
```

Reply:
Noticed that getOrPut doesn't allow null value to be returned by defaultValue function ,
so in places where it is possible to receive null after computing for the key , we can use ?.let scope and putIfAbsent to account for NullPointerException.
Have implemented it accordingly.

Reply:
Ah I see. I want to try and avoid the synchronized bit if we can since it's pretty highly discouraged in Kotlin (see https://kt.academy/article/ek-synchronization) since it can cause problems with coroutines (see https://blog.danlew.net/2020/01/28/coroutines-and-java-synchronization-dont-mix/).

That being said, getOrPut does not guarantee atomicity, unlike does computeIfAbsent for ConcurrentHashMap. This may actually be a case where a slight change in implementation is needed to better leverage a lockless approach via coroutines. I'll follow up in a second comment with suggested versions of this & the next class to try and better illustrate what I'm thinking.

Reply:
Here's what I'm thinking for fixing the issue with locale creation: https://gist.github.com/BenHenning/deb77757e89e1974066a6e76d1496f8b/revisions. There will be some peripheral changes needed in other files, too.

For FakeAssetRepository, we ideally would update it in the same way (i.e. update AssetRepository to returned ``Deferred``s instead, which would also let us get rid of the ReentrantLock AssetRepositoryImpl is currently using). However, that's a change that will affect a lot more in the codebase, and doesn't seem worth it for the immediate benefit of fixing the crash. Could we maybe instead copy the ReentrantLock pattern used in the production implementation (https://github.com/oppia/oppia-android/blob/develop/utility/src/main/java/org/oppia/android/util/caching/AssetRepositoryImpl.kt)? It has the same problem as synchronized inherently, though it offers better consistency with other locks used in the codebase and offers a variety of benefits over the built-in JVM lock (see https://stackoverflow.com/a/11821900).